File: 2 SEGUNDO/1 Cuatrimestre/Programación y Diseño Orientado a Objetos - PDOO/Practicas/rubyNapakalaki/player.rb

require_relative 'treasure'
require_relative 'bad_consequence.rb'
require_relative 'dice.rb'
require_relative 'combat_result.rb'


class Player
  attr_accessor :name, :level, :dead, :can_i_steal, :hidden_treasures, :visible_treasures, :pending_bad_consequence ,:enemy
  protected :enemy

  @@CONST_MAXLEVEL = 10


   def initialize(n,d,l,v_t,h_t,cis,pbc)

      @name = n
      @dead = d
      @level = l
      @visible_treasures = v_t
      @hidden_treasures = h_t
      @can_i_steal = cis
      @pending_bad_consequence = pbc


   end

   def self.copia(p)

     self.new(p.name,p.dead,p.level,p.visible_treasures,p.hidden_treasures,p.can_i_steal,p.pending_bad_consequence)

   end

   def is_death
      @dead
    end


   def should_convert

      dice = Dice.get_instance.next_number
      sol = false

     if dice == 1
         sol = true
     end
      sol
   end

   protected :should_convert



   def  get_oponent_level(m)
      m.get_combatlevel

   end

     protected :get_oponent_level

  def bring_to_life

     @dead = false;

  end

  private :bring_to_life

  def get_level_treasures

    res  = 0

    visible_treasures.each do |vis|

      res = res + vis.bonus

    end

    res

  end

  private :get_level_treasures


   def get_combat_level


     combat_level=0;

     combat_level = get_level_treasures

        combat_level = combat_level + @level;

    combat_level

   end

   protected :get_combat_level


   def increment_levels(i)

     if( @level + i >= @@CONST_MAXLEVEL )
        @level = @@CONST_MAXLEVEL

     else

       @level = @level + i

     end
   end

   private :increment_levels

    def decrement_levels(i)

       if(@level-i <= 1)
          @level = 1

       else
        @level = @level - i

      end

    end

    private :decrement_levels


  def set_pending_bad_consequence(b)

     @pending_bad_consequence = b

  end

  private :set_pending_bad_consequence


  def die_if_no_treasures

     if @hidden_treasures.length	 == 0 and @visible_treasures.length == 0

         @dead = true
     end

  end


  def valid_state

     res=true

     if ((!@pending_bad_consequence.is_empty) or @hidden_treasures.length > 4)

         res = false
     end

     res
  end

  def how_many_visible_treasures(t_kind)

      res = 0


    @visible_treasures.each do |vis|

      if vis.type == t_kind then

        res = res + 1

      end

    end


    res

  end

  private :how_many_visible_treasures

  def set_enemy_player(enemy)

      @enemy = enemy
  end


  def have_stolen

      @can_i_steal = false

  end

  private :have_stolen

  def can_you_give_me_a_treasure

    res = false

      if @hidden_treasures.length > 0  then

          res = true

      end

    res

  end


    def can_make_treasure_visible(t)

       result = false

      case t.get_type()

      when TreasureKind::ONEHAND

        if is_treasure_kind_in_use(TreasureKind::BOTHHANDS) then
          result = false
        else

          i = 0
          @visible_treasures.each do |tv|
            if tv.get_type() == TreasureKind::ONEHAND then
              i += 1
            end

          end

          if i == 2 then
            result = false
          else
            result = true
          end
        end

      else
        result = !is_treasure_kind_in_use(t.get_type())

      end

      result
    end

    private :can_make_treasure_visible

    def is_treasure_kind_in_use(type)

      result = false
      @visible_treasures.each do |tesoro|
        if type == tesoro.get_type() then

          result = true
          break

        end

      end
      result
    end
    private :is_treasure_kind_in_use


  def apply_prize(m)

    increment_levels(m.get_levels_gained)
    tesoros = m.get_treasures_gained

    baraja = CardDealer.get_instance

        for i in 0..tesoros-1
          @hidden_treasures << baraja.next_treasure

        end
   end

  private :apply_prize

    def apply_bad_consequence(m)

      bad = m.get_bad_consequence

      if bad.is_dead == false
                puts bad.class.to_s
                pending_bad = bad.adjust_to_fit_treasure_list(@visible_treasures,@hidden_treasures)

                niveles = pending_bad.getLevels

                decrement_levels(niveles)


                pending_bad.setLevels(0)

                @pending_bad_consequence = pending_bad

      else if  bad.is_dead == true






          decrement_levels(@@CONST_MAXLEVEL)


          discardAllTreasures

           puts 'Al morir el jugador, este pierde todos los tesoros de los que dispone
                                (tanto equipados como ocultos) y su nivel queda fijado en 1.'
       end

      end

    end

  private :apply_bad_consequence

  def stealTreasure
    can_i = canISteal

    if(can_i)
      can_you = @enemy.can_you_give_me_a_treasure

      if(can_you)
        treasure = enemy.give_me_a_treasure
        @hidden_treasures << treasure
        @enemy.hidden_treasures.delete(treasure)
        @enemy.die_if_no_treasures

       have_stolen

      end
    end
    treasure
  end

  def  give_me_a_treasure

    maximo = @hidden_treasures.length - 1

    pos_aleatorio = Random.rand(0..maximo)

    res = @hidden_treasures.at(pos_aleatorio)
    @hidden_treasures.delete(res)

   res
  end

  def discardVisibleTreasure(t)

      @visible_treasures.delete(t)

      if @pending_bad_consequence != nil  and !@pending_bad_consequence.is_empty

          @pending_bad_consequence.substract_visible_treasure(t)

      if @pending_bad_consequence.getnVisibleTreasures > 0 then
              n = @pending_bad_consequence.getnVisibleTreasures
               n=n-1
              @pending_bad_consequence.setnVisibleTreasures(n)
          end
      end
      die_if_no_treasures
      CardDealer.get_instance.give_treasure_back(t)


  end



  def discardHiddenTreasure(t)

     @hidden_treasures.delete(t)

      if @pending_bad_consequence != nil  and !@pending_bad_consequence.is_empty

          @pending_bad_consequence.substract_hidden_treasure(t)

      if @pending_bad_consequence.getnHiddenTreasures > 0 then
              n = @pending_bad_consequence.getnHiddenTreasures
               n=n-1
              @pending_bad_consequence.setnHiddenTreasures(n)
          end
      end
      die_if_no_treasures
      CardDealer.get_instance.give_treasure_back(t)


  end


  def discardAllTreasures

    visible = @visible_treasures.clone
    oculto = @hidden_treasures.clone
    for i in visible
      discardVisibleTreasure(i)
    end
    for j in oculto
      discardHiddenTreasure(j)
    end



    die_if_no_treasures
  end


  def init_treasures

      dealer = CardDealer.instance
      d = Dice.instance
      bring_to_life

      treasure = dealer.next_treasure
      @hidden_treasures << treasure

      number = d.next_number

      if number > 1
          treasure = dealer.next_treasure
          @hidden_treasures << treasure
      end

      if number == 6
          treasure = dealer.next_treasure();
          @hidden_treasures << treasure

      end



  end

  def combat(m)


      my_level = get_combat_level
      monster_level = m.get_combatlevel


      if my_level > monster_level

        apply_prize(m)

          if @level >= @@CONST_MAXLEVEL
            combate = NapakalakiGame::CombatResult::WINGAME

          else
            combate = NapakalakiGame::CombatResult::WIN
          end

      else
          apply_bad_consequence(m)
          die_if_no_treasures
          if should_convert
            combate = NapakalakiGame::CombatResult::LOSEANDCONVERT

          else
            combate = NapakalakiGame::CombatResult::LOSE
          end

      end

      combate
  end

  def canISteal
    @can_i_steal
  end

  def makeTreasureVisible(t)

      can_i = can_make_treasure_visible(t)

      if can_i
          @visible_treasures << t
          @hidden_treasures.delete(t);
      end
  end

  def getVisibleTreasures
    @visible_treasures
  end

  def getHiddenTreasures
   @hidden_treasures
  end

  def to_s
    "#{@name} \nLevel: #{@level}\n"
  end

end

File: 2 SEGUNDO/1 Cuatrimestre/Programación y Diseño Orientado a Objetos - PDOO/Practicas/javaNapakalaki/NapakalakiGame/Player.java
package NapakalakiGame;

import java.util.ArrayList;
import java.util.Arrays;



public class Player {

 private static final int MAXLEVEL = 10;
 private String name;
 protected int level;
 private boolean dead = true;
 private boolean canISteal = true;
 private ArrayList<Treasure> hiddenTreasures = new ArrayList<Treasure>();
 private ArrayList<Treasure> visibleTreasures = new ArrayList<Treasure>();
 protected Player enemy;
 private BadConsequence pendingBadConsequence = new NumericBadConsequence("",0,0,0);
 private boolean isCultist = false;
 private boolean texto = false;


 public Player (String nombre){
     name = nombre;
     level = 1 ;
 }

 public Player (Player p){

   name = p.name;
   level = p.level;
   dead = p.dead;
   canISteal = p.canISteal;
   hiddenTreasures = (ArrayList<Treasure>) (p.hiddenTreasures).clone();
   visibleTreasures =  (ArrayList<Treasure>) (p.visibleTreasures).clone();
   enemy = p.enemy;
   pendingBadConsequence = p.pendingBadConsequence;

 }

 protected int getOponentLevel(Monster m ){
     return m.getCombatLevel();
 }

 protected boolean shouldConvert(){
     int dado = (int)(Math.random() * 5 + 1);
     boolean sol = false;
     if(dado == 1){
         sol = true;
     }
     return sol;
 }


 public boolean isDeath(){

    boolean death = false;
    if(dead){
        death = true;
    }
    return death;
 }

   @Override
     public String toString(){
       
        String res = " "+ name +" Nivel: " + level+  "\n"
                + "Mal rollo a cumplir: " + pendingBadConsequence + "\n";
        return res ;
      }

 public String getName(){
     return name;
 }


 public Player getEnemy(){
     return enemy;
 }

 public BadConsequence getPendingBadConsequence(){
     return pendingBadConsequence;
 }

 public ArrayList<Treasure> getVisibleTreasures(){
     return visibleTreasures;
 }

 public ArrayList<Treasure> getHiddenTreasures(){
     return hiddenTreasures;
 }

 private void bringToLife(){
     dead = false;
 }

 protected int getLevelTreasures(){
   int res  = 0;
   for (Treasure iterador: visibleTreasures){
      res = res+ iterador.getBonus();
    }
    return res;
 }

 protected int getCombatLevel(){
    int combatLevel = getLevelTreasures();
    combatLevel = combatLevel + level;

    return combatLevel;
 }

 private void incrementLevels(int i){

     if(level+i >= MAXLEVEL){
        level = MAXLEVEL;
     }else{
        level = level + i;
     }
 }
 private void decrementLevels(int i){

     if(level-i <= 1){
        level = 1;
     }else{
        level = level - i;
     }
 }

 private void setPendingBadConsequence(BadConsequence b){
    pendingBadConsequence = b;
 }

 private void dieIfNoTreasures(){
   if(hiddenTreasures.isEmpty()  && visibleTreasures.isEmpty()){
     dead = true;
    }
 }

 public boolean validState(){

     boolean res=true;

     if(!(pendingBadConsequence.isEmpty()) || hiddenTreasures.size() > 4){
       res = false;
     }
     return res;
 }

  public int getLevel(){
     return level;
 }

  protected int howManyVisibleTreasures(TreasureKind tKind){
    int res = 0;

   for (Treasure iterador: visibleTreasures){
      if(iterador.getType() == tKind){
        res++;
      }
    }
     return res;
  }

  protected int howManyHiddenTreasures(TreasureKind tKind){
    int res = 0;
     for (Treasure iterador: visibleTreasures){
      if(iterador.getType() == tKind){
        res++;
      }
    }
   return res;
  }

  public void setEnemyPlayer(Player enemy){
      this.enemy = enemy;
  }

  public boolean canISteal(){
      return canISteal;
  }

  private void haveStolen(){
      canISteal = false;
  }

  private boolean canYouGiveMeATreasure(){
      boolean res = false;

      if(hiddenTreasures.size()> 0){
          res = true;
      }
    return res;
  }


  public boolean canMakeTreasureVisible(Treasure t){

    boolean res=true;

    if(visibleTreasures.size()<4){

    TreasureKind elemento = t.getType();

    switch (elemento){

    case ONEHAND:

        if (howManyVisibleTreasures(TreasureKind.BOTHHANDS)>0) {
            res = false;
        } else {
          int i = 0;
          for (Treasure tv : this.visibleTreasures) {
              if (tv.getType() == (TreasureKind.ONEHAND)) {
                  i++;
              }
          }
          if (i == 2) {
              res = false;
          } else {
              res = true;
          }
        }
        break;

    case BOTHHANDS:
      if(howManyVisibleTreasures(TreasureKind.ONEHAND)>0){
          res = false;
      } else {
        int j = 0;
        for (Treasure tv : this.visibleTreasures) {
            if (tv.getType() == (TreasureKind.BOTHHANDS)) {
                j++;
            }
        }

        if (j == 1) {
            res = false;
        } else {
            res = true;
        }
      }
      break;

    default:
        res = howManyVisibleTreasures(elemento)==0;
        break;
          }

      }else{
          res = false;
      }
  return res;
  }

  private void applyPrize(Monster m){
      this.incrementLevels(m.getLevelsGained());
      int tesoros = m.getTreasuresGained();

      CardDealer baraja = CardDealer.getInstance();

      for(int i = 0; i < tesoros; i++){
          hiddenTreasures.add(baraja.nextTreasure());
      }
  }

  private void applyBadConsequence(Monster m){

    BadConsequence bad = m.getBadConsequence();
    BadConsequence pendingBad;
    boolean actualizar = false;

    if(!(bad.isDeath())){

      pendingBad = bad.adjustToFitTreasureLists(visibleTreasures,hiddenTreasures);

      int niveles = pendingBad.getLevels();
      this.decrementLevels(niveles);

      pendingBad.setLevels(0);
      this.setPendingBadConsequence(pendingBad);
      muerto(actualizar);

    }else if(bad.isDeath()){
      actualizar = true;
      muerto(actualizar);
      this.decrementLevels(MAXLEVEL);
      this.discardAllTreasures();
     }
  }

  public boolean muerto(boolean muerte){

      if(muerte == true){
          texto = true;
      }else{
          texto = false;
      }
      return texto;
  }

  public boolean getT(){
      return texto;
  }

  public Treasure stealTreasure(){
      boolean canI = this.canISteal();
      Treasure treasure = null;

      if(canI){
          boolean canYou = enemy.canYouGiveMeATreasure();
          if(canYou){
              treasure=enemy.giveMeATreasure();
              this.hiddenTreasures.add(treasure);
              enemy.hiddenTreasures.remove(treasure);
              enemy.dieIfNoTreasures();
              this.haveStolen();
          }
      }
      return treasure;
  }

  private Treasure giveMeATreasure(){

       Treasure res;
       int posAleatorio;
       posAleatorio = (int) (Math.random() * hiddenTreasures.size());
       res = hiddenTreasures.get(posAleatorio);

       return res;
  }

  public void discardVisibleTreasure(Treasure t){

      visibleTreasures.remove(t);

      if((pendingBadConsequence != null) && (!pendingBadConsequence.isEmpty())){
            pendingBadConsequence.substractVisibleTreasure(t);
           if(pendingBadConsequence.getnVisibleTreasures() > 0){
            int n = pendingBadConsequence.getnVisibleTreasures();
            n--;
            pendingBadConsequence.setnVisibleTreasures(n);
           }
      }

      dieIfNoTreasures();
 }

  public void discardHiddenTreasure(Treasure t){

      hiddenTreasures.remove(t);

      if((pendingBadConsequence != null) && (!pendingBadConsequence.isEmpty())){
            pendingBadConsequence.substractHiddenTreasure(t);
        if(pendingBadConsequence.getnHiddenTreasures() > 0){
            int n = pendingBadConsequence.getnHiddenTreasures();
            n--;
            pendingBadConsequence.setnHiddenTreasures(n);

            }
      }

      dieIfNoTreasures();
  }

  public void discardAllTreasures(){
    ArrayList<Treasure> visible = new ArrayList();
    ArrayList<Treasure> oculto = new ArrayList();

    for(int i = 0; i< visibleTreasures.size(); i++){
        visible.add(visibleTreasures.get(i));
    }

     for(int i = 0; i< visible.size(); i++){
         this.discardVisibleTreasure(visible.get(i));
    }

    for(int i = 0; i< hiddenTreasures.size(); i++){
        oculto.add(hiddenTreasures.get(i));
    }

     for(int i = 0; i< oculto.size(); i++){
         this.discardHiddenTreasure(oculto.get(i));
    }
    dieIfNoTreasures();
  }

  public void initTreasures(){
    CardDealer dealer = CardDealer.getInstance();

    Dice dice = Dice.getInstance();

    this.bringToLife();

    Treasure treasure = dealer.nextTreasure();
    this.hiddenTreasures.add(treasure);

    int number = (int)(Math.random() * 5 + 1);


    if (number > 1){
        treasure = dealer.nextTreasure();
        this.hiddenTreasures.add(treasure);
    }else
    if (number == 6) {
      treasure = dealer.nextTreasure();
      this.hiddenTreasures.add(treasure);
    }
  }

  public CombatResult combat(Monster m){
      CombatResult combatResult;
      int myLevel = getCombatLevel();
      int monsterLevel = this.getOponentLevel(m);

      if(myLevel > monsterLevel){
          this.applyPrize(m);

          if(level >=MAXLEVEL){
                combatResult = CombatResult.WINGAME;
          }
          else{
               combatResult = CombatResult.WIN;
          }
      }else{
        this.applyBadConsequence(m);
        dieIfNoTreasures();

        if(this.shouldConvert()){
           isCultist = true;
           combatResult = CombatResult.LOSEANDCONVERT;

        }else{
          combatResult = CombatResult.LOSE;
        }
      }
    return combatResult;
  }

  public void makeTreasureVisible(Treasure t){

      boolean canI = canMakeTreasureVisible(t);

      if(canI){
          visibleTreasures.add(t);
          hiddenTreasures.remove(t);
      }
  }

  public boolean isCultist(){
    return isCultist;
  }

}
